### PR TITLE
feat: add haunted support

### LINKED ISSUE #54: 🎃 Haunted Support 👻
Apollo Elements should support [haunted](https://github.com/matthewp/haunted) with `useQuery`, `useMutation` hooks

### PR DESCRIPTION
**Pull request by bennypowers**
closes #54

### CODE REVIEW

File: packages/haunted/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  - [👾 Mutations](#-mutations)
+  - [🗞 Subscriptions](#-subscriptions)
+- [👷‍♂️ Maintainers](#-maintainers)
+
+## 🔧 Installation
+
+Apollo Elements haunted hooks are distributed through `npm`, the node package manager. To install a copy of the latest version in your project's `node_modules` directory, [install npm on your system](https://www.npmjs.com/get-npm) then run the following command in your project's root directory:
+
+```bash
+npm install --save @apollo-elements/haunted
```
Comment (matthewp) [inline packages/haunted/README.md:23]: 🙏

File: packages/haunted/useQuery.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      called: undefined,
+      startPolling: (ms: number) =>
+        void (this.pollingInterval = window.setInterval(refetch, ms)),
+      stopPolling: () =>
+        clearInterval(this.pollingInterval),
+    };
+  }
+
+  teardown() {
+    this.state.host.disconnectedCallback();
```
Comment (matthewp) [inline packages/haunted/useQuery.ts:132]: Is this for testing? Usually you wouldn't call disconnectedCallback in an element.

Comment (bennypowers) [reply packages/haunted/useQuery.ts:132]: generally speaking, `connectedCallback` does three things:
1. set up a mutation observer
2. initialize `document` and `variables` from DOM content, if applicable
3. fire an `apollo-element-connected` event, which `<apollo-client>` listens for

https://github.com/apollo-elements/apollo-elements/blob/3e546195fcfa2016ad395494fe79491cbdb9ac70/packages/mixins/apollo-element-mixin.ts#L102-L113

and the `disconnectedCallback` does one thing:
1. fire an `apollo-element-disconnected` event, which `<apollo-client>` listens for.

https://github.com/apollo-elements/apollo-elements/blob/3e546195fcfa2016ad395494fe79491cbdb9ac70/packages/mixins/apollo-element-mixin.ts#L118-L122

Ideally, this behaviour will be uniform across all implementations, and defined only once (i.e. in the mixin)

**Comment by bennypowers** [comment]
Ok we've got some decent progress here. I've solved all the major bugs with `useQuery`. I still need to implement `updateQuery` prop and review the docs, but looks good so far.

Next up is mutation and subscription

**Comment by bennypowers** [comment]
Alright now we have a version of `useMutation` that passes component integration tests 💪

File: packages/haunted/useQuery.ts (hunk 2 above)
Comment (matthewp) [reply packages/haunted/useQuery.ts:156]: Ok, this was just an observation of mine, don't think you're doing anything wrong necessarily.

**Comment by matthewp** [comment]
Looks great! Let me know when this is released and I'll tweet / retweet about it

**Comment by bennypowers** [comment]
Just about ready to go, once an [issue with the test runner](https://github.com/modernweb-dev/web/issues/993) is resolved.

**Comment by bennypowers** [comment]
Alright I think we're ready to rock. I'm planning to publish later tonight, tweet tomorrow, and then maybe add a live example later on.

**Comment by matthewp** [comment]
Thanks for this @bennypowers !

**Pull request merged**